chain/chain/src/store.rs

```diff
@@ -1817,26 +1817,49 @@ impl<'a> ChainStoreUpdate<'a> {
             .insert((*hash, shard_id), outgoing_receipts);
     }
 
+    /// For all the outgoing receipts generated in block `hash` at the shards we are tracking
+    /// in this epoch,
+    /// save a mapping from receipt ids to the destination shard ids that the receipt will be sent
+    /// to in the next block.
+    /// Note that this function should be called after `save_block` is called on this block because
+    /// it requires that the block info is available in EpochManager.
     pub fn save_receipt_id_to_shard_id(
         &mut self,
+        me: &Option<AccountId>,
         runtime_adapter: &dyn RuntimeAdapter,
+        hash: &CryptoHash,
         prev_hash: &CryptoHash,
-        shard_id: ShardId,
-        last_height_included: BlockHeight,
+        num_shards: NumShards,
     ) -> Result<(), Error> {
-        let outgoing_receipts = self.chain_store.get_outgoing_receipts_for_shard(
-            runtime_adapter,
-            *prev_hash,
-            shard_id,
-            last_height_included,
-        )?;
-        let shard_layout = runtime_adapter.get_shard_layout_from_prev_block(prev_hash)?;
-        for receipt in outgoing_receipts {
-            let to_shard_id = account_id_to_shard_id(&receipt.receiver_id, &shard_layout);
-            self.chain_store_cache_update
-                .receipt_id_to_shard_id
-                .insert(receipt.receipt_id, to_shard_id);
+        for shard_id in 0..num_shards {
+            if runtime_adapter.cares_about_shard(me.as_ref(), &prev_hash, shard_id as ShardId, true)
+            {
+                let receipt_id_to_shard_id: HashMap<_, _> = {
+                    // it can be empty if there is no new chunk for this shard
+                    if let Ok(outgoing_receipts) = self.get_outgoing_receipts(hash, shard_id) {
```

Comment:
I see that you've changed from get_outgoing_receipts_for_shard to get_outgoing_receipts.
(I assume this is intentional - as the difference is quite subtle)

Reply:
Yes that's intentional